### PR TITLE
[GlobalOptimization] Create linalg.generic ops instead of CastOpInter…

### LINKED ISSUE #15547: [CPU][DT] ExpandVector hoists arith op outside of generic op
This issue leads to a compilation error when DT is enabled:

Repro:

```
iree-opt test.mlir --iree-global-opt-expand-vectors

func.func @main(%157 : tensor<256128xi1>, %49 : tensor<256128x1536xf32>) -> tensor<1536xf32> {
  %c0 = arith.constant 0.0 : f32
  %c22 = arith.constant dense<22.0> : tensor<1536xf32>
  %158 = tensor.empty() : tensor<256128xf32>
  %159 = linalg.generic {indexing_maps = [affine_map<(d0) -> (d0)>, affine_map<(d0) -> (d0)>], iterator_types = ["parallel"]} ins(%157 : tensor<256128xi1>) outs(%158 : tensor<256128xf32>) {
  ^bb0(%in: i1, %out: f32):
    %1923 = arith.uitofp %in : i1 to f32
    linalg.yield %1923 : f32
  } -> tensor<256128xf32>
  %160 = tensor.empty() : tensor<1536xf32>
  %161 = linalg.fill ins(%c0 : f32) outs(%160 : tensor<1536xf32>) -> tensor<1536xf32>
  %162 = linalg.vecmat ins(%159, %49 : tensor<256128xf32>, tensor<256128x1536xf32>) outs(%161 : tensor<1536xf32>) -> tensor<1536xf32>
  %163 = tensor.empty() : tensor<1x1x1536xf32>
  %164 = linalg.generic {indexing_maps = [affine_map<(d0) -> (d0)>, affine_map<(d0) -> (d0)>, affine_map<(d0) -> (d0)>], iterator_types = ["parallel"]} ins(%162, %c22 : tensor<1536xf32>, tensor<1536xf32>) outs(%160 : tensor<1536xf32>) {
  ^bb0(%in: f32, %in_794: f32, %out: f32):
    %1923 = arith.mulf %in, %in_794 : f32
    linalg.yield %1923 : f32
  } -> tensor<1536xf32>
  return %164 : tensor<1536xf32>
}
```

Output:
```
#map = affine_map<(d0) -> (d0)>
module {
  func.func @main(%arg0: tensor<256128xi1>, %arg1: tensor<256128x1536xf32>) -> tensor<1536xf32> {
    %cst = arith.constant 0.000000e+00 : f32
    %cst_0 = arith.constant dense<2.200000e+01> : tensor<1536xf32>
    %0 = tensor.empty() : tensor<1536xf32>
    %1 = linalg.fill ins(%cst : f32) outs(%0 : tensor<1536xf32>) -> tensor<1536xf32>
    %expanded = tensor.expand_shape %arg0 [[0, 1]] : tensor<256128xi1> into tensor<1x256128xi1>
    %2 = arith.uitofp %expanded : tensor<1x256128xi1> to tensor<1x256128xf32>
    %expanded_1 = tensor.expand_shape %1 [[0, 1]] : tensor<1536xf32> into tensor<1x1536xf32>
    %3 = linalg.matmul ins(%2, %arg1 : tensor<1x256128xf32>, tensor<256128x1536xf32>) outs(%expanded_1 : tensor<1x1536xf32>) -> tensor<1x1536xf32>
    %collapsed = tensor.collapse_shape %3 [[0, 1]] : tensor<1x1536xf32> into tensor<1536xf32>
    %4 = linalg.generic {indexing_maps = [#map, #map, #map], iterator_types = ["parallel"]} ins(%collapsed, %cst_0 : tensor<1536xf32>, tensor<1536xf32>) outs(%0 : tensor<1536xf32>) {
    ^bb0(%in: f32, %in_2: f32, %out: f32):
      %5 = arith.mulf %in, %in_2 : f32
      linalg.yield %5 : f32
    } -> tensor<1536xf32>
    return %4 : tensor<1536xf32>
  }
}
```

See how `arith.uitofp` is no longer within a generic op after running this pass, which is incorrect.

### PR DESCRIPTION
**Pull request by Max191**
…face ops on tensors

Closes #15547

### CODE REVIEW

**Review by NatashaKnk — APPROVED** [review]
Thanks for the fix! 
@dcaballe this should fix the compilation error

File: compiler/src/iree/compiler/GlobalOptimization/Utils.cpp [hunk 1]
```diff
@@ -52,6 +53,39 @@ std::optional<Type> getCastElemType(Value input) {
...
 }
 
+FailureOr<Value>
+createGenericElementwiseCastOp(OpBuilder &builder, Location loc, Value input,
+                               CastOpInterface castOp,
+                               ArrayRef<NamedAttribute> attrs) {
+  auto inputType = dyn_cast<RankedTensorType>(input.getType());
+  if (!inputType) {
+    return failure();
+  }
```
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/Utils.cpp]: It is very wrong if we still have unranked tensor at this stage. Can we remove the check or just add an `assert`? In this context, we can return `Value` type directly.

File: compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp [hunk 2]
```diff
@@ -164,8 +164,9 @@ static Value castEncodedResult(OpBuilder &builder, Location loc, Value encoded,
   SmallVector<utils::IteratorType> iteratorTypes(castedType.getRank(),
                                                  utils::IteratorType::parallel);
   auto genericOp = castOp->getParentOfType<linalg::GenericOp>();
-  NamedAttrList castAttrs =
-      genericOp ? genericOp->getAttrs() : castOp->getAttrs();
+  NamedAttrList castAttrs = genericOp
+                                ? linalg::getPrunedAttributeList(genericOp)
+                                : castOp->getAttrs();
```
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp:152]: Can we use `createGenericElementwiseCastOp` method here?

File: compiler/src/iree/compiler/GlobalOptimization/Utils.h [hunk 3]
```diff
@@ -7,27 +7,39 @@
 #define IREE_COMPILER_GLOBALOPTIMIZATION_UTILS_H_
 
 #include <optional>
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
```
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/Utils.h]: This is not needed for header file right? 

I don't see a requirement for this after skimming through the `Utils.cpp` file. If it is needed, can you move it to `Utils.cpp`?

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/GlobalOptimization/Utils.h (hunk 3 above)
Comment (Max191) [reply compiler/src/iree/compiler/GlobalOptimization/Utils.h]: My VSCode must have automatically added that. I can replace both of those includes with just `mlir/Support/LogicalResult.h`

**Review by hanhanW — APPROVED** [review]

**Pull request merged**